### PR TITLE
Support for comments on table in all databases

### PR DESCRIPTION
**Pull request by moufmouf**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no

#### Summary

I opened a previous PR to add support for table level comments on PostgreSQL: #3499 

This PR is an attempt to add the same feature on all databases (and not only MySQL / PostgreSQL).

This PR is built on top of #3499 that already adds support to PostgreSQL.

Commit https://github.com/doctrine/dbal/commit/c4d83a7ca3a6abdfb78610a5bc64ce76ca65e182 moves the functional test from PostgreSQL to the abstract functional test case. Obviously, this test will fail as long as all databases do not support table comments.

Closes #3499

### CODE REVIEW

**Comment by moufmouf** [comment]
Commit https://github.com/doctrine/dbal/pull/3512/commits/fb9d2b3f28b4853842fd1d6755b268e92a228d07 adds support for table comments in Sqlite.

Those databases still need some love:

- [ ] Oracle
- [ ] SQL Server
- [ ] DB2

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (greg0ire) [inline lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: Since we require php 7.2, you can remove this phpdoc and use type hints instead:
```suggestion
    public function getCommentOnTableSQL(string $tableName, ?string $comment): string
```

Comment (greg0ire) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: Also, should this method be `final`?

File: lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php
Comment (greg0ire) [inline lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php]: This pattern is quite long… maybe we should use the x modifier here.

>  It is also possible to use bracket style delimiters where the opening and closing brackets are the starting and ending delimiter, respectively. (), {}, [] and <> are all valid bracket style delimiter pairs. 

TIL :bulb: , this is nice!

Comment (greg0ire) [reply lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php]: Here is how it could look:
```
        $pattern = '{[\s]*CREATE TABLE(?:\W"' . preg_quote($this->_platform->quoteSingleIdentifier($table)) . '"\W|\W' . preg_quote($table). '\W)
(
    ?:\(.*?\)
    |
    [^,(]
)
*?,?(
    (?:(?!\n)) # comment explaining what this is
    (?:\s*--[^\n]*\n?)+ # other comment, etc.
)}ix';
```

Comment (moufmouf) [reply lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php]: Excellent idea. Actually, spending some time commenting the regular expression allowed me to spot some useless parts. Done in https://github.com/doctrine/dbal/pull/3512/commits/3e381ef151a78d63c5aa7624f4e0a00e24cb8341

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: Done. I added type-hints and removed PHPDoc. Moving to PHP 7.2+ rocks :)

> Also, should this method be final? Same question for all the methods you introduced.

I reviewed all added methods and think they should not be final. Most of the methods in `AbstractPlatform` are meant to be overridden in sub-classes if some database requires a special syntax.

**Comment by moufmouf** [comment]
Current status: thanks to this PR, DB comments are now working on SQLServer and SQlite.
I've "skipped"  tests on Oracle and DB2 so this PR can be merged (I can open a PR later for Oracle and DB2)

**Comment by moufmouf** [comment]
.... and support in DB2 and Oracle has been added! \o/

**Comment by moufmouf** [comment]
Since all databases now support table level comments, I took the liberty to add a `Table::getComment` and a `Table::setComment` method.
In order to avoid causing BC breaks with MySQL user that previously relied on the "comment" option, those methods are simply writing a "comment" option.

We can remove the "comment" option in the next major release.

File: lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php
Comment (greg0ire) [reply lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php]: Lovely!

File: lib/Doctrine/DBAL/Platforms/DB2Platform.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/DB2Platform.php]: Please do not change the variable semantic once it's declared. The original meaning is the value itself, the modified meaning is the SQL expression representing the value. Something similar to the following would be preferred (including the nowdoc):

https://github.com/doctrine/dbal/blob/9853ad0c0ed2d5a3f12704c01c405fc820d06438/lib/Doctrine/DBAL/Platforms/MySqlPlatform.php#L376-L384

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: It's questionable that comments are nullable. There doesn't seem to be a difference between NULL and an empty string in this case. Would it make sense to not allow NULL in the first place to avoid a potential BC break? It may become irrelevant if the method is not exposed as `public`.

File: lib/Doctrine/DBAL/Platforms/OraclePlatform.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/OraclePlatform.php]: Please use `quoteStringLiteral()` immediately during concatenation and leave the variable unchanged. Otherwise, it's not reusable (see the comment above).

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: Not really informative. Please remove.

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: Is this method really needed?

File: lib/Doctrine/DBAL/Schema/DB2SchemaManager.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/DB2SchemaManager.php]: There seems to be some repetition of the code like this: most platforms override `listTableDetails()`, call the parent and add the comment. Probably, this logic should be part of the parent method itself (most of the platforms support comments).

File: lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php]: Is it possible that the method will return NULL? It should always return an array.

File: lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php]: `empty()` cannot be used here. This way, a comment like '`0`' will be ignored.

**Review by morozov — COMMENTED** [review]
@moufmouf thank you for putting this together. We may need to do a couple of rounds of reviews and changes to work through the newly introduced APIs.

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: It is a good question. I actually tried to mimic what was done with the comments on columns to be consistent.

There is already a `public function getCommentOnColumnSQL()` method: https://github.com/doctrine/dbal/blob/4ea7c119d32d7fc1f7755cb5464a2f4f34e60fa4/lib/Doctrine/DBAL/Platforms/AbstractPlatform.php#L1639-L1656

If I was to start from zero, I would probably make the method protected, put it in a trait, and "use" this trait directly in the XXXSchemaManager classes. But since there is already a `getCommentOnColumnSQL` method, I thought it would be better to do something similar.

I'm really open to any change here, if you think there is a better way.

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: Completely agree, but same thing here. I simply copied the `getCommentOnColumnSQL` method behaviour. In `getCommentOnColumnSQL`, comments are already a nullable string. Shall I change this or keep this as-is for consistency?

File: lib/Doctrine/DBAL/Platforms/DB2Platform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/DB2Platform.php]: > We select most of the table details in a single SQL.

Actually, there is one query to fetch all table names, and then, per table, there is one query for all the columns, one for all the indexes and one for the foreign keys.

https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php#L245-L277

But the first query used to list the table names returns only the names.

When there are additional table metadata to be fetched (as in the MySQLSchemaManager), the table comments are fetched in the same query as the other metadata: https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Platforms/MySqlPlatform.php#L383-L395
You can notice that in the current implementation, there is already N+1 queries to fetch MySQL metadata (including table comments).

We may change `listTableNames` into something like `listTableNamesAndComments` or `listTableNamesAndMetadata` that would return the table names AND the comments at once. What do you think?

File: lib/Doctrine/DBAL/Platforms/DB2Platform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/DB2Platform.php]: :+1: done

File: lib/Doctrine/DBAL/Platforms/OraclePlatform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/OraclePlatform.php]: :+1: done

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: :+1: Done

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: Mmmm... well... :grin: I mostly copied what is going on in the MySQLPlatform: https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Platforms/MySqlPlatform.php#L522

Actually, I cannot think of any good reason why we would want to do this. I'll remove this in SqlitePlatform. Should I remove this also in MySQLPlatform?

File: lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php]: :+1: good catch

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: It is useful to have it as a separate method because the statement can be different (typically, this method is overloaded for SQLServer). However, you are right that it is not necessary to have this "public". I'm moving this to "protected".

File: lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php]: The signature of the `getCreateTableSQL` method says otherwise: https://github.com/doctrine/dbal/blob/master/lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php#L483 :disappointed:

File: lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php]: :+1: changed

File: lib/Doctrine/DBAL/Schema/DB2SchemaManager.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Schema/DB2SchemaManager.php]: True, but there is a catch.

For most DBMS we want to fetch the comments. For MySQL, we fetch the comments + a bunch of metadata.

What we could do is design a method that returns an array of "options" to be applied to the table. Something like:

```php
    public function listTableDetails($tableName)
    {
        // ....
        $options      = $this->getTableOptions($tableName);
        foreach ($options as $key => $value) {
            $table->addOption($key, $value);
        }
        return $table;
    }
```

Now, for each `XXXSchemaManager`, we can provide a protected `getTableOptions` method that is specific to the schema manager. Would that be ok?

**Comment by moufmouf** [comment]
@morozov thanks a lot for the review. I took in account most of your comments and added some questions/comments where I'm not completely sure what the best solution is.

File: lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php
Comment (greg0ire) [inline lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php]: How about using nowdoc here, instead of escaping all this single quotes?

File: lib/Doctrine/DBAL/Schema/OracleSchemaManager.php
Comment (greg0ire) [inline lib/Doctrine/DBAL/Schema/OracleSchemaManager.php]: This method should probably have at least a return value, correct?

Comment (moufmouf) [reply lib/Doctrine/DBAL/Schema/OracleSchemaManager.php]: Absolutely :+1:

File: lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/SQLServerPlatform.php]: Done :+1:

File: lib/Doctrine/DBAL/Schema/OracleSchemaManager.php
Comment (greg0ire) [reply lib/Doctrine/DBAL/Schema/OracleSchemaManager.php]: Please do the same for all methods below then :P

Comment (moufmouf) [reply lib/Doctrine/DBAL/Schema/OracleSchemaManager.php]: Ooops, my bad :)
Done in https://github.com/doctrine/dbal/pull/3512/commits/fc053b1e1a6ceeba6954eedef5269f5cbb93d03d

**Comment by morozov** [comment]
@moufmouf sorry, I don't have enough time for this issue right now, so please don't take my absence for ignoring. Unless other core team members help you finish the code changes, I'll try to get back to this as soon as I can.

File: lib/Doctrine/DBAL/Schema/MySqlSchemaManager.php
Comment (SenseException) [inline lib/Doctrine/DBAL/Schema/MySqlSchemaManager.php]: If master is going to be DBAL 2.10, adding a return type would be a BC break.

**Review by SenseException — DISMISSED** [review]

File: lib/Doctrine/DBAL/Schema/MySqlSchemaManager.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Schema/MySqlSchemaManager.php]: Indeed! Solved.

**Comment by moufmouf** [comment]
Any more feedback on this?
Ping @morozov

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (Ocramius) [reply lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: I think it should be removed, unless we're breaking a test highlighting why that's there

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (Ocramius) [reply lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: Still requiring feedback from @moufmouf

**Review by Ocramius — DISMISSED** [review]
LGTM :+1:

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: The code to generate a SQL comment for a column or for a table is the same in `SqlitePlatform`.

I could directly call `$this->getInlineColumnCommentSQL($comment)` from here: 

https://github.com/doctrine/dbal/blob/4ea7c119d32d7fc1f7755cb5464a2f4f34e60fa4/lib/Doctrine/DBAL/Platforms/SqlitePlatform.php#L337-L342

But of course, we would use the "getInlineColumnCommentSQL" to retrieve a table comment and this seems weird. Creating an additional `getInlineTableCommentSQL` method makes it clear this is the behaviour that was intended.

I'm happy to remove the `getInlineTableCommentSQL`  method and directly point to `getInlineColumnCommentSQL` instead if you prefer. You tell me.

File: lib/Doctrine/DBAL/Platforms/SqlitePlatform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/SqlitePlatform.php]: @morozov @Ocramius I removed this useless trim. It does not break a test and I still cannot think of why this was put here in the first place. Probably dead code from another DBMS that made it through to the SqlitePlatform. Good riddance :)

**Comment by moufmouf** [comment]
Trying to revive this stalled PR (sorry, I got a lot of work lately).
I took into account @Ocramius comments regarding the useless trimming of table comments in Sqlite.

I pulled from master in order to get the latest changes and the build is now failing for a reason that seems unrelated to this PR.
@Ocramius  @morozov Do you see any more changes to be made?

File: lib/Doctrine/DBAL/Platforms/AbstractPlatform.php
Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/AbstractPlatform.php]: Let's keep it consistent with the rest of the methods for now.

File: lib/Doctrine/DBAL/Platforms/DB2Platform.php
Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/DB2Platform.php]: Please use permalinks when referring to code fragments. This way it's guaranteed that the same fragment will be referred regardless of the future code changes.

Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/DB2Platform.php]: Some platforms like Oracle build schema introspection views on the fly and it's very inefficient to build them N times instead of just once. We should avoid introducing an API which will knowingly have performance problems.

I'd say a comment is part of the metadata/details by its meaning. API-wise, why would we want to have a separate method which returns only table names and their comments?

Would it make sense to introduce something like `AbstractPlatform::getListTablesWithDetails()`? Then `AbstractSchemaManager::listTables()`, instead of getting a list of tables and fetching their details one by one would call the new method.

File: lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php
Comment (morozov) [reply lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php]: Please use permalinks instead of `master`. The link is already outdated.

**Comment by morozov** [comment]
@moufmouf please rebase. All existing issues should be resolved, however MySQL 8 builds may still fail.

File: lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Schema/SqliteSchemaManager.php]: Edit previous comment with permalink. Sorry for that. I just learned about the 'y' keyboard shortcut to get those permalinks. :)

File: lib/Doctrine/DBAL/Platforms/DB2Platform.php
Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/DB2Platform.php]: > Some platforms like Oracle build schema introspection views on the fly and it's very inefficient to build them N times instead of just once. We should avoid introducing an API which will knowingly have performance problems.
>
> I'd say a comment is part of the metadata/details by its meaning. API-wise, why would we want to have a separate method which returns only table names and their comments?

I wholeheartedly agree.

> Would it make sense to introduce something like `AbstractPlatform::getListTablesWithDetails()`? Then `AbstractSchemaManager::listTables()`, instead of getting a list of tables and fetching their details one by one would call the new method.

I'll give this a try.

Comment (moufmouf) [reply lib/Doctrine/DBAL/Platforms/DB2Platform.php]: @morozov I started working on this, but I ran into an issue.

Currently, the `listTables` method calls `listTablesDetails` in a loop. This is logical because we need to fetch all columns of all tables.

https://github.com/doctrine/dbal/blob/29486f06cae54166f19423b6bf87bab73fb58ec1/lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php#L245-L260

The instantiation of the `Table` object is done in `listTablesDetails`:

https://github.com/doctrine/dbal/blob/29486f06cae54166f19423b6bf87bab73fb58ec1/lib/Doctrine/DBAL/Schema/AbstractSchemaManager.php#L267-L277

In my current attempt, `listTables` looks like this (in pseudo-code):

```
function listTables() {
    $tableRows = platform->getTableListWithMetadata()

    foreach ($tableRows as $tableRow) {
        $table = $this->listTableDetails($tableRow['name']);

        $table->setComment($tableRow['comment'])
        $tables[] = $table;
    }
    return $tables;
}
```

This works perfectly well... except that if a user calls `listTablesDetails` directly (instead of going through `listTables`, the table object returned will not contain the metadata.

This means I should consider writing 2 variants of `listTablesDetails`. One that can be called directly and that actively fetches the metadata. And one that is provided the metadata.
I can certainly do that is you wish, but it adds some complexity to the code. Do you want me to keep working in that direction or is the additional complexity not worth performance benefit?

**Comment by moufmouf** [comment]
@morozov Rebase done.

File: lib/Doctrine/DBAL/Platforms/DB2Platform.php
Comment (morozov) [reply lib/Doctrine/DBAL/Platforms/DB2Platform.php]: Yeah… effort-wise, this is getting out of control. This is solvable in the way that there should be a method which returns all columns of all tables so it gets called only once for all tables but it's really out of scope.

Let's leave it as is and thank you for giving it a try.

**Comment by moufmouf** [comment]
@morozov @Ocramius I think I handled all comments on this PR.

Is there anything else you want me to do or should I let it rest here until it is merged?

(more specifically, OracleSchemaManager is now rated C by Scrutinizer after I added a new method in it, but I don't really see how to make that rating go away)

**Comment by morozov** [comment]
> Is there anything else you want me to do or should I let it rest here until it is merged?

I think we're good. Please squash the commits unless you think some of them are relevant fro the history.

**Comment by moufmouf** [comment]
@morozov Squash done.
Thanks a lot for taking the time to review this PR!

**Review by morozov — APPROVED** [review]

**Comment by morozov** [comment]
:ship:

**Pull request merged**
